### DISCUSSION TITLE
Traits cannot be set during self-service OIDC registration

### DISCUSSION THREAD

**Discussion by openscript** [ory/kratos#3413] [Q&A]
I've created the following identity traits:

```
{
  "$id": "https://myproject.com/identity.schema.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Customer",
  "type": "object",
  "properties": {
    "traits": {
      "type": "object",
      "properties": {
        "lang": {
          "type": "string",
          "title": "Language"
        },
        "tac": {
          "type": "string",
          "title": "Terms and Conditions"
        },
        "dp": {
          "type": "string",
          "title": "Data Protection"
        },
        "email": {
          "type": "string",
          "format": "email",
          "title": "Email",
          "minLength": 3,
          "ory.sh/kratos": {
            "credentials": {
              "password": {
                "identifier": true
              }
            },
            "verification": {
              "via": "email"
            },
            "recovery": {
              "via": "email"
            }
          }
        }
      },
      "required": [
        "email"
      ],
      "additionalProperties": false
    }
  }
}
```

They work well of password self-service flows. When I make a GET Request to `.ory/kratos/public/self-service/registration` I receive:

```
{
  "id": "46448ab4-fa72-4fd1-922e-af35278bca71",
  "type": "api",
  "expires_at": "2023-08-01T13:36:55.811222326Z",
  "issued_at": "2023-08-01T13:06:55.811222326Z",
  "request_url": "http://myproject.com/self-service/registration/api?return_session_token_exchange_code=true&return_to=myproject%3A%2F%2Fregistration",
  "return_to": "myproject://registration",
  "ui": {
    "action": "https://myproject.com/.ory/kratos/public/self-service/registration?flow=46448ab4-fa72-4fd1-922e-af35278bca71",
    "method": "POST",
    "nodes": [
      {
        "type": "input",
        "group": "default",
        "attributes": {
          "name": "csrf_token",
          "type": "hidden",
          "value": "GTXbhBHZWLlDNg0lgCIsIzK8A4lxQOWOySFT1m/kFd6xxvvgTb1hwfgaBOcwOxN12brJ/n8K/G5cpPcAuLuzoQ==",
          "required": true,
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {}
      },
      {
        "type": "input",
        "group": "oidc",
        "attributes": {
          "name": "provider",
          "type": "submit",
          "value": "apple",
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {
          "label": {
            "id": 1040002,
            "text": "Sign up with apple",
            "type": "info",
            "context": {
              "provider": "apple"
            }
          }
        }
      },
      {
        "type": "input",
        "group": "oidc",
        "attributes": {
          "name": "provider",
          "type": "submit",
          "value": "google",
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {
          "label": {
            "id": 1040002,
            "text": "Sign up with google",
            "type": "info",
            "context": {
              "provider": "google"
            }
          }
        }
      },
      {
        "type": "input",
        "group": "password",
        "attributes": {
          "name": "traits.lang",
          "type": "text",
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {
          "label": {
            "id": 1070002,
            "text": "Language",
            "type": "info"
          }
        }
      },
      {
        "type": "input",
        "group": "password",
        "attributes": {
          "name": "password",
          "type": "password",
          "required": true,
          "autocomplete": "new-password",
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {
          "label": {
            "id": 1070001,
            "text": "Password",
            "type": "info"
          }
        }
      },
      {
        "type": "input",
        "group": "password",
        "attributes": {
          "name": "traits.tac",
          "type": "text",
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {
          "label": {
            "id": 1070002,
            "text": "Terms and Conditions",
            "type": "info"
          }
        }
      },
      {
        "type": "input",
        "group": "password",
        "attributes": {
          "name": "traits.dp",
          "type": "text",
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {
          "label": {
            "id": 1070002,
            "text": "Data Protection",
            "type": "info"
          }
        }
      },
      {
        "type": "input",
        "group": "password",
        "attributes": {
          "name": "traits.email",
          "type": "email",
          "required": true,
          "autocomplete": "email",
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {
          "label": {
            "id": 1070002,
            "text": "Email",
            "type": "info"
          }
        }
      },
      {
        "type": "input",
        "group": "password",
        "attributes": {
          "name": "method",
          "type": "submit",
          "value": "password",
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {
          "label": {
            "id": 1040001,
            "text": "Sign up",
            "type": "info",
            "context": {}
          }
        }
      }
    ]
  },
  "session_token_exchange_code": "Fox5xcz9g3C37reHYAhFlHk0px8EaysZV2d5WTECtfooLIGdj3STu5jk65SgC2ha"
}
```

As the traits are only in the group `password` they are not available for `oidc` registration. Is this correct? How to set the traits for OIDC account?

**Comment by openscript** [comment]
Any ideas? I'm also happy to pay for the support, but I'm not a big tech corp :)

**Comment by jonas-jonas** [comment]
Hi! Kratos assumes, that the traits all come from the OIDC provider. You can use the [Jsonnet data mapping](https://www.ory.sh/docs/kratos/social-signin/data-mapping) to achieve that.

If that's not the case, you'll need to add a custom form inside your application after the registration to collect the data (e.g. via a `return_to` in the registration flow). That's a pattern that some applications use to collect additional data during the registration process that's not available from the OIDC provider, such as shipping addresses, etc.

In your case, you might be able to get the `lang` from the provider (depending on the provider). As for T&C and data processing, that depends on your requirements, but making these "checked" by default for OIDC registrations can also be achieved in the Jsonnet mapping. 

```jsonnet
local claims = std.extVar('claims');
{
  identity: {
    traits: {
      // The email might be empty if the user hasn't granted permissions for the email scope.
      [if 'email' in claims then 'email' else null]: claims.email, // adjust for your provider's mapping
      tac: true,
      dp: true,
      lang: // depends on your provider...
    },
  },
}
```

_(untested, but should work)_

**Reply by openscript** [reply]
Thank you very much for your response! I've been wondering this, because with `0.12` of Kratos it was possible to add the traits already during the registration also for OIDC flows. I checked the logs for this breaking change, but didn't find anything. The API docs still have the traits mentioned:
![image](https://github.com/ory/kratos/assets/1105080/481f3c00-24d6-4c0b-8f93-e516128fbef4)


Anyway, it's also fine to collect this information after registration succeeded as these traits don't come from the OIDC. We are legally required to store the date when a user accepted the terms & conditions as well as the data protection.

The idea is to use the [updateSettingsFlow](https://www.ory.sh/docs/kratos/reference/api#tag/frontend/operation/updateSettingsFlow) to set the traits post registration, right?

**Reply by jonas-jonas** [reply]
>because with 0.12 of Kratos it was possible to add the traits already during the registration also for OIDC flows

Hm, I don't remember anything changing in that area, at least not deliberately. 

>The idea is to use the [updateSettingsFlow](https://www.ory.sh/docs/kratos/reference/api#tag/frontend/operation/updateSettingsFlow) to set the traits post registration, right?

Yes. Either that, or if you really just need the date, that might be possible using the data mapping or a [post registration hook](https://www.ory.sh/docs/guides/integrate-with-ory-cloud-through-webhooks#update-identity-traits), that returns the updated fields.

**Reply by openscript** [reply]
> Hm, I don't remember anything changing in that area, at least not deliberately.

As I still see it in the documentation, it should be possible to set traits during OIDC registration? I've been looking for a test about this case, but didn't find any. I'm happy to open an issue and try to write a failing test if you agree.

**Reply by jonas-jonas** [reply]
>As I still see it in the documentation, it should be possible to set traits during OIDC registration?

Not actually sure about that. As I said, Kratos assumes that traits come from the OIDC provider, and the other form is a different strategy ("password" vs "oidc"). 

And the fact that it's included in the docs, just means that the endpoint allows a `traits` field but doesn't say that it's actually used. This is because the endpoint is used for all strategies and due to the documentation being generated from the code.

**Reply by openscript** [reply]
When I first encountered this issue, I also thought it's because of the different strategy, but it was working with Kratos 0.12 for almost 9 months.

The code looks like it should work with 1.0 too, as the traits from the requests are handed over to the continuity container here: https://github.com/ory/kratos/blob/3a07af4d4ffd8372139bd6e1a3063088244e8bd6/selfservice/strategy/oidc/strategy_registration.go#L165-L170

Also the OIDC specific struct `UpdateRegistrationFlowWithOidcMethod` where the documentation is generated from, contains `traits`.

**Comment by carlosTijerino** [comment]
Hello, I have a similar problem with oidc and traits parameters with google provider, how did you manage to solve it? @openscript

**Reply by openscript** [reply]
It's a bug also within Orys SaaS, that haven't been resolved for more than half a year: https://github.com/ory/network/issues/345

Unfortunately I don't know a workaround.

**Reply by carlosTijerino** [reply]
Thanks for your answer
